### PR TITLE
Frontend metrics

### LINKED ISSUE #9522: Metrics tracker for Builder
Create frontend tracker to send events in the builder to the back.

- [x] Created analysis
- [x] Modified analysis
- [x] Removed analysis
- [x] Created map
- [x] Deleted map
- [x] Exported map
- [x] Published map
- [x] Created dataset
- [x] Deleted dataset

**Branch**: frontend-metrics

### PR DESCRIPTION
**Pull request by MariaCheca**
Closes #9522 

Added new module `MetricsTracker` to communicate events occurring in the front to the metrics API. Each event (i.e. create analysis) is preceded by a `MetricsTracker.track()` call to create a new `MetricsModel` object which sends the data we need to the back.

Front CR @nobuti please :) cc @gfiorav 

**_Frontend tests passing in dev environment despite those Travis failed builds._**

### CODE REVIEW

**Comment by nobuti** [comment]
@MariaCheca I see some unchecked points in the issue, are they still pending?

**Comment by MariaCheca** [comment]
hey @nobuti! Actually, that checklist was made when we started the feature but is not accurate. I removed it from the issue to avoid misunderstandings and added the current ones :)

File: lib/assets/javascripts/cartodb/common/dialogs/delete_layer/delete_layer_view.js [hunk 1]
```diff
@@ -66,6 +67,8 @@ module.exports = BaseDialog.extend({
       wait: true,
       success: function() {
         self.close();
+
+        MetricsTracker.track('Deleted dataset', this.attributes);
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb/common/dialogs/delete_layer/delete_layer_view.js]: Mmm weird. This deletes a layer, the dataset still remains in the user's profile, right? Not sure if we should change the event name.

File: lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js [hunk 2]
```diff
@@ -79,6 +80,8 @@ module.exports = CoreView.extend({
       self._visDefinitionModel.set('visChanges', 0);
       self._mapcapsCollection.add(data, {at: 0});
       self.model.set({status: 'updated'});
+
+      MetricsTracker.track('Published map', self._visDefinitionModel.attributes);
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js]: This is tricky, this button is the same for publish but for update too. We should take account of this, something like this:
```
var eventName = self._mapcapsCollection.length > 0 ? 'Updated map' : 'Published map';
MetricsTracker.track(eventName, self._visDefinitionModel.attributes);
```

File: lib/assets/javascripts/cartodb3/editor/layers/operations/remove-layer.js [hunk 3]
```diff
@@ -21,6 +22,8 @@ module.exports = function (opts) {
         info: _t('editor.layers.delete.success', {name: name}),
         closable: true
       });
+
+      MetricsTracker.track('Deleted dataset', layerDefinitionModel.attributes);
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/layers/operations/remove-layer.js]: Same concern as above

File: lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js (hunk 2 above)
Comment (MariaCheca) [inline lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js]: That was a good one and I didn't notice, thank you!

File: lib/assets/javascripts/cartodb/common/dialogs/delete_layer/delete_layer_view.js (hunk 1 above)
Comment (MariaCheca) [inline lib/assets/javascripts/cartodb/common/dialogs/delete_layer/delete_layer_view.js]: Totally right, fixing that.

File: lib/assets/javascripts/cartodb/common/dialogs/create/create_map_model.js [hunk 4]
```diff
@@ -3,6 +3,7 @@ var _ = require('underscore-cdb-v3');
 var ImportsModel = require('../../background_polling/models/imports_model');
 var UploadModel = require('../../background_polling/models/upload_model');
 var VisFetchModel = require('../../visualizations_fetch_model');
+var MetricsTracker = require('../../../../cartodb3/components/metrics/metrics-tracker');
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb/common/dialogs/create/create_map_model.js]: Are we sure this is working at all? This code belongs to the old editor. We are doing `MetricsTracker.init` in the builder, in `editor.js`, but at this point, this object doesn't exist. Also some objects like configModel, doesn't exist there.

File: lib/assets/javascripts/cartodb/common/dialogs/create/listing/navigation_view.js [hunk 5]
```diff
@@ -2,6 +2,7 @@ var cdb = require('cartodb.js-v3');
 var _ = require('underscore-cdb-v3');
 var Utils = require('cdb.Utils');
 var pluralizeString = require('../../../view_helpers/pluralize_string');
+var MetricsTracker = require('../../../../../cartodb3/components/metrics/metrics-tracker');
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb/common/dialogs/create/listing/navigation_view.js]: Same as above.

**Comment by nobuti** [comment]
The code for builder is nice, but not sure if this is going to work when the `MetricsTracker` is attached to the editor's code. Could we test it in staging to make sure is working fine?

File: lib/assets/javascripts/cartodb/common/dialogs/create/create_map_model.js (hunk 4 above)
Comment (MariaCheca) [inline lib/assets/javascripts/cartodb/common/dialogs/create/create_map_model.js]: Nope, this wasn't working, we were receiving the event from other part inside the builder which is correctly tracking the creation of maps. As we want this trackers only in the builder, I'm removing this call and the one below.

File: lib/assets/javascripts/cartodb3/dataset/dataset-header/header-view.js [hunk 6]
```diff
@@ -197,7 +198,11 @@ module.exports = CoreView.extend({
         },
         runAction: function () {
           tableDeleteOperation({
-            onSuccess: self._onSuccessDestroyDataset.bind(self, modalModel),
+            onSuccess: function () {
+              self._onSuccessDestroyDataset.bind(self, modalModel);
+
+              MetricsTracker.track('Deleted dataset', self._tableModel.attributes);
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/dataset/dataset-header/header-view.js]: 👍  but we should add the `MetricsTracker.init` to the `dataset.js` file in the same way we do in `editor.js`.

Comment (MariaCheca) [inline lib/assets/javascripts/cartodb3/dataset/dataset-header/header-view.js]: After talking with @gfiorav this is not necessary anymore as this is present in the back too (back has more priority than front if the event tracker can be set there).

**Comment by MariaCheca** [comment]
retest this please

**Comment by MariaCheca** [comment]
✨

File: lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js [hunk 7]
```diff
@@ -186,6 +187,8 @@ module.exports = CoreView.extend({
   _onCreateEmptyClick: function () {
     if (this._userModel.canCreateDatasets()) {
       this._createModel.createFromScratch();
+
+      MetricsTracker.track('Created dataset', this._createModel.attributes);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js]: This is not necessary, we should have done it in the backend cc @gfiorav.

File: lib/assets/javascripts/cartodb3/data/export-map-definition-model.js [hunk 8]
```diff
@@ -38,6 +39,8 @@ module.exports = Backbone.Model.extend({
         error: this._errorHandler.bind(this)
       });
     }.bind(this), 2000);
+
+    MetricsTracker.track('Exported map', this.attributes);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/data/export-map-definition-model.js]: cc @MariaCheca

File: lib/assets/javascripts/cartodb3/editor/map-operations/remove-map.js [hunk 9]
```diff
@@ -13,6 +14,8 @@ module.exports = function (opts) {
     wait: true,
     success: function (mdl, attrs) {
       successCallback && successCallback(mdl, attrs);
+
+      MetricsTracker.track('Removed map', visDefinitionModel.attributes);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/map-operations/remove-map.js]: Remove it @MariaCheca

File: lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js (hunk 7 above)
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js]: @MariaCheca 

could you provide a list of attributes that are sent here? 

That could help us purge out the duplicate events.

File: lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js [hunk 10]
```diff
@@ -79,6 +80,10 @@ module.exports = CoreView.extend({
       self._visDefinitionModel.set('visChanges', 0);
       self._mapcapsCollection.add(data, {at: 0});
       self.model.set({status: 'updated'});
+
+      if (self._mapcapsCollection.length < 0) {
+        MetricsTracker.track('Published map', self._visDefinitionModel.attributes);
```
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/components/modals/publish/publish-button-view.js]: Same here

File: lib/assets/javascripts/cartodb3/data/export-map-definition-model.js (hunk 8 above)
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/data/export-map-definition-model.js]: and here

File: lib/assets/javascripts/cartodb3/editor/map-operations/remove-map.js (hunk 9 above)
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/editor/map-operations/remove-map.js]: This one you can ignore: there is no event called "Removed map", this call returns an error.

File: lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js (hunk 7 above)
Comment (gfiorav) [inline lib/assets/javascripts/cartodb3/components/modals/add-layer/content/navigation-view.js]: Scratch that, this event doesn't seem to we working anyway, so no problem.

**Pull request merged**